chain/tsresolver/tipset_resolver.go

```diff
@@ -0,0 +1,164 @@
+//   - a 0x-prefixed hex block number: the TipSet at the given height
+//
+// If a specific block number is specified and `strict` is true, an error is returned if the block
+// number resolves to a null round, otherwise in the case of a null round the first non-null TipSet
+// immediately before the null round is returned.
+func (tsr *tipSetResolver) ResolveEthBlockSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	switch selector {
+	case EthBlockSelectorEarliest:
+		return nil, fmt.Errorf(`block param "%s" is not supported`, EthBlockSelectorEarliest)
+
+	case EthBlockSelectorPending:
+		return tsr.loader.GetHeaviestTipSet(), nil
+
+	case EthBlockSelectorLatest:
+		// head - 1 because we're always one behind for Eth compatibility due to deferred execution
+		return tsr.loader.LoadTipSet(ctx, tsr.loader.GetHeaviestTipSet().Parents())
+
+	case EthBlockSelectorSafe, EthBlockSelectorFinalized:
+		defaultDelay := policy.ChainFinality
+		if selector == EthBlockSelectorSafe {
+			defaultDelay = ethtypes.SafeEpochDelay
+		}
+
+		head := tsr.loader.GetHeaviestTipSet()
+		latestHeight := head.Height() - 1 // always one behind for Eth compatibility due to deferred execution
+		defaultHeight := latestHeight - defaultDelay
+
+		if f3TipSet, err := tsr.getF3FinalizedTipSet(ctx); err != nil {
+			return nil, err
+		} else if f3TipSet != nil && f3TipSet.Height() > defaultHeight {
+			// return the parent of the finalized tipset (deferred execution, eth cares about t)
+			return tsr.loader.LoadTipSet(ctx, f3TipSet.Parents())
+		} // else F3 is disabled, not running, or behind the default safe or finalized height
+
+		ts, err := tsr.loader.GetTipsetByHeight(ctx, defaultHeight, head, true)
+		if err != nil {
+			return nil, xerrors.Errorf("loading tipset at height %v: %w", defaultHeight, err)
+		}
+		return ts, nil
+
+	default:
+		// likely an 0x hex block number or a decimal block number
+		return tsr.resolveEthBlockNumberSelector(ctx, selector, strict)
+	}
+}
+
+func (tsr *tipSetResolver) getF3FinalizedTipSet(ctx context.Context) (*types.TipSet, error) {
+	if !tsr.f3.IsEnabled() {
+		return nil, nil
+	}
+	if running, _ := tsr.f3.IsRunning(); !running {
+		return nil, nil
+	}
+
+	cert, err := tsr.f3.GetLatestCert(ctx)
+	if err != nil {
+		log.Debugf("loading latest F3 certificate: %s", err)
+		return nil, nil
+	}
+
+	tsk, err := types.TipSetKeyFromBytes(cert.ECChain.Head().Key)
+	if err != nil {
+		return nil, xerrors.Errorf("decoding tipset key reported by F3: %w", err)
+	}
+
+	finalizedTipSet, err := tsr.loader.LoadTipSet(ctx, tsk)
+	if err != nil {
+		return nil, xerrors.Errorf("loading tipset reported as finalized by F3 %s: %w", tsk, err)
+	}
+
+	return finalizedTipSet, nil
+}
+
+func (tsr *tipSetResolver) resolveEthBlockNumberSelector(ctx context.Context, selector string, strict bool) (*types.TipSet, error) {
+	var num ethtypes.EthUint64
+	if err := num.UnmarshalJSON([]byte(`"` + selector + `"`)); err != nil {
```

Comment:
most of the hard work for this function happen in `EthUint64#UnmarshalJSON`: https://github.com/filecoin-project/lotus/blob/dbf41383b37f1155dd0a28b10da2909226adb171/chain/types/ethtypes/eth_types.go#L51-L75